modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RegisteredDomainProcessor.java

```diff
@@ -0,0 +1,186 @@
+
+    private final String field;
+    private final String targetField;
+    private final String targetETLDField;
+    private final String targetSubdomainField;
+    private final boolean ignoreMissing;
+
+    RegisteredDomainProcessor(
+        String tag,
+        String description,
+        String field,
+        String targetField,
+        String targetETLDField,
+        String targetSubdomainField,
+        boolean ignoreMissing
+    ) {
+        super(tag, description);
+        this.field = field;
+        this.targetField = targetField;
+        this.targetETLDField = targetETLDField;
+        this.targetSubdomainField = targetSubdomainField;
+        this.ignoreMissing = ignoreMissing;
+    }
+
+    public String getField() {
+        return field;
+    }
+
+    public String getTargetField() {
+        return targetField;
+    }
+
+    public String getTargetETLDField() {
+        return targetETLDField;
+    }
+
+    public String getTargetSubdomainField() {
+        return targetSubdomainField;
+    }
+
+    public boolean getIgnoreMissing() {
+        return ignoreMissing;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+        DomainInfo info = getRegisteredDomain(ingestDocument);
+        if (info == null) {
+            if (ignoreMissing) {
+                return ingestDocument;
+            } else {
+                throw new IllegalArgumentException("unable to set domain information for document");
+            }
+        }
+        if (info.getRegisteredDomain() != null) {
+            ingestDocument.setFieldValue(targetField, info.getRegisteredDomain());
+        }
+        if (targetETLDField != null && info.getTLD() != null) {
+            ingestDocument.setFieldValue(targetETLDField, info.getTLD());
+        }
+        if (targetSubdomainField != null && info.getSubdomain() != null) {
+            ingestDocument.setFieldValue(targetSubdomainField, info.getSubdomain());
+        }
+        return ingestDocument;
+    }
+
+    private DomainInfo getRegisteredDomain(IngestDocument d) {
+        String fieldString = d.getFieldValue(field, String.class, ignoreMissing);
+        if (fieldString == null) {
+            return null;
+        }
+        String registeredDomain = SUFFIX_MATCHER.getDomainRoot(fieldString);
+        if (registeredDomain == null) {
+            if (SUFFIX_MATCHER.matches(fieldString)) {
+                return new DomainInfo(fieldString);
+            }
+            return null;
+        }
+        if (registeredDomain.indexOf(".") == -1) {
+            // we have domain with no matching public suffix, but "." in it
+            return null;
+        }
+        return new DomainInfo(registeredDomain, fieldString);
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    private class DomainInfo {
+        private final String registeredDomain;
+        private final String tld;
```

Comment:
This is eTLD, right?

Reply:
yes, I can rename it to be clearer if you'd prefer.